File: backend/app/api/scripts.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy import func, select
from sqlalchemy.orm import Session

from app.core.database import get_db
from app.models.scene import Scene
from app.models.scene_script import SceneScript
from app.models.script import Script
from app.models.script_version import ScriptVersion
from app.schemas.script import (
    ScriptCreate,
    ScriptRead,
    ScriptSceneReference,
    ScriptUpdate,
    ScriptValidateRequest,
    ScriptValidateResponse,
)
from app.services.yaml_validator import validate_yaml_content

router = APIRouter(prefix="/api/scripts", tags=["scripts"])


def _script_scene_references(db: Session, script_id: int) -> list[ScriptSceneReference]:
    scenes = list(
        db.scalars(
            select(Scene).join(SceneScript, SceneScript.scene_id == Scene.id).where(SceneScript.script_id == script_id)
        ).all()
    )
    return [ScriptSceneReference(id=scene.id, name=scene.name) for scene in scenes]


@router.get("", response_model=list[ScriptRead])
def list_scripts(db: Session = Depends(get_db)) -> list[ScriptRead]:
    stmt = (
        select(Script, func.count(SceneScript.id).label("scene_count"))
        .outerjoin(SceneScript, SceneScript.script_id == Script.id)
        .group_by(Script.id)
        .order_by(Script.updated_at.desc())
    )
    return [
        ScriptRead.model_validate(
            {
                "id": script.id,
                "name": script.name,
                "content": script.content,
                "source_type": script.source_type,
                "created_at": script.created_at,
                "updated_at": script.updated_at,
                "scene_count": scene_count,
                "scenes": [],
            }
        )
        for script, scene_count in db.execute(stmt).all()
    ]


@router.get("/{script_id}", response_model=ScriptRead)
def get_script(script_id: int, db: Session = Depends(get_db)) -> ScriptRead:
    script = db.get(Script, script_id)
    if not script:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Script not found")
    scene_count = db.scalar(select(func.count(SceneScript.id)).where(SceneScript.script_id == script_id)) or 0
    return ScriptRead.model_validate(
        {
            "id": script.id,
            "name": script.name,
            "content": script.content,
            "source_type": script.source_type,
            "created_at": script.created_at,
            "updated_at": script.updated_at,
            "scene_count": scene_count,
            "scenes": _script_scene_references(db, script_id),
        }
    )


@router.post("", response_model=ScriptRead, status_code=status.HTTP_201_CREATED)
def create_script(payload: ScriptCreate, db: Session = Depends(get_db)) -> ScriptRead:
    script = Script(name=payload.name, content=payload.content, source_type=payload.source_type)
    db.add(script)
    db.flush()

    version = ScriptVersion(script_id=script.id, version_no=1, content=script.content)
    db.add(version)

    db.commit()
    db.refresh(script)
    return ScriptRead.model_validate(
        {
            "id": script.id,
            "name": script.name,
            "content": script.content,
            "source_type": script.source_type,
            "created_at": script.created_at,
            "updated_at": script.updated_at,
            "scene_count": 0,
            "scenes": [],
        }
    )


@router.put("/{script_id}", response_model=ScriptRead)
def update_script(script_id: int, payload: ScriptUpdate, db: Session = Depends(get_db)) -> ScriptRead:
    script = db.get(Script, script_id)
    if not script:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Script not found")

    if payload.name is not None:
        script.name = payload.name
    if payload.content is not None:
        script.content = payload.content
    if payload.source_type is not None:
        script.source_type = payload.source_type

    current_version = db.scalar(
        select(func.coalesce(func.max(ScriptVersion.version_no), 0)).where(ScriptVersion.script_id == script_id)
    )
    new_version = ScriptVersion(
        script_id=script.id,
        version_no=(current_version or 0) + 1,
        content=script.content,
    )
    db.add(new_version)

    db.commit()
    db.refresh(script)
    scene_count = db.scalar(select(func.count(SceneScript.id)).where(SceneScript.script_id == script_id)) or 0
    return ScriptRead.model_validate(
        {
            "id": script.id,
            "name": script.name,
            "content": script.content,
            "source_type": script.source_type,
            "created_at": script.created_at,
            "updated_at": script.updated_at,
            "scene_count": scene_count,
            "scenes": _script_scene_references(db, script_id),
        }
    )


@router.delete("/{script_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_script(script_id: int, db: Session = Depends(get_db)) -> None:
    script = db.get(Script, script_id)
    if not script:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Script not found")
    db.delete(script)
    db.commit()


@router.post("/{script_id}/copy", response_model=ScriptRead, status_code=status.HTTP_201_CREATED)
def copy_script(script_id: int, db: Session = Depends(get_db)) -> ScriptRead:
    origin = db.get(Script, script_id)
    if not origin:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Script not found")

    copied = Script(
        name=f"{origin.name}-copy",
        content=origin.content,
        source_type=origin.source_type,
    )
    db.add(copied)
    db.flush()

    db.add(ScriptVersion(script_id=copied.id, version_no=1, content=copied.content))
    db.commit()
    db.refresh(copied)
    return ScriptRead.model_validate(
        {
            "id": copied.id,
            "name": copied.name,
            "content": copied.content,
            "source_type": copied.source_type,
            "created_at": copied.created_at,
            "updated_at": copied.updated_at,
            "scene_count": 0,
            "scenes": [],
        }
    )


@router.post("/{script_id}/validate", response_model=ScriptValidateResponse)
def validate_script(
    script_id: int,
    payload: ScriptValidateRequest,
    db: Session = Depends(get_db),
) -> ScriptValidateResponse:
    if not db.get(Script, script_id):
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Script not found")

    result = validate_yaml_content(payload.content)
    return ScriptValidateResponse(
        valid=result.valid,
        line=result.line,
        column=result.column,
        message=result.message,
    )

File: backend-ts/src/routes/scripts.ts
import type { FastifyInstance } from "fastify";

import { env } from "../config.js";
import { prisma } from "../db/prisma.js";
import { badRequest, notFound, sendError } from "../lib/errors.js";
import { serializeDate } from "../lib/serializers.js";
import { AndroidExecutionError, executeCompiledScene } from "../modules/midscene/index.js";
import {
  compileSceneScript,
  dumpTaskSnapshot,
  extractScriptEnv,
  parseScriptTasks,
  SceneCompileError
} from "../services/scene-compiler.js";
import { scriptCreateSchema, scriptUpdateSchema } from "../schemas.js";

function serializeScriptRead(script: {
  id: number;
  name: string;
  content: string;
  source_type: string;
  created_at: Date;
  updated_at: Date;
  scene_count?: number;
  scenes?: Array<{ id: number; name: string }>;
}) {
  return {
    id: script.id,
    name: script.name,
    content: script.content,
    source_type: script.source_type,
    created_at: serializeDate(script.created_at),
    updated_at: serializeDate(script.updated_at),
    scene_count: script.scene_count ?? 0,
    scenes: script.scenes ?? []
  };
}

async function scriptSceneReferences(scriptId: number) {
  return prisma.scene.findMany({
    where: {
      sceneScripts: {
        some: {
          script_id: scriptId
        }
      }
    },
    select: {
      id: true,
      name: true
    }
  });
}

export async function registerScriptRoutes(app: FastifyInstance): Promise<void> {
  app.get("/api/scripts", async (_request, reply) => {
    try {
      const scripts = await prisma.script.findMany({
        orderBy: {
          updated_at: "desc"
        },
        include: {
          _count: {
            select: {
              sceneScripts: true
            }
          }
        }
      });
      return scripts.map((script) =>
        serializeScriptRead({
          ...script,
          scene_count: script._count.sceneScripts,
          scenes: []
        })
      );
    } catch (error) {
      return sendError(reply, error);
    }
  });

  app.get("/api/scripts/:scriptId", async (request, reply) => {
    try {
      const scriptId = Number((request.params as { scriptId: string }).scriptId);
      const script = await prisma.script.findUnique({ where: { id: scriptId } });
      if (!script) {
        notFound("Script not found");
      }
      const [sceneCount, scenes] = await Promise.all([
        prisma.sceneScript.count({ where: { script_id: scriptId } }),
        scriptSceneReferences(scriptId)
      ]);
      return serializeScriptRead({
        ...script,
        scene_count: sceneCount,
        scenes
      });
    } catch (error) {
      return sendError(reply, error);
    }
  });

  app.get("/api/scripts/:scriptId/tasks", async (request, reply) => {
    try {
      const scriptId = Number((request.params as { scriptId: string }).scriptId);
      const script = await prisma.script.findUnique({ where: { id: scriptId } });
      if (!script) {
        notFound("Script not found");
      }
      return parseScriptTasks(script.content).map((item) => ({
        script_id: script.id,
        task_index: item.task_index,
        task_name: item.task_name,
        continue_on_error: item.continue_on_error,
        task_content: dumpTaskSnapshot(item.task)
      }));
    } catch (error) {
      if (error instanceof SceneCompileError) {
        return reply.code(400).send({ detail: error.message });
      }
      return sendError(reply, error);
    }
  });

  app.post("/api/scripts", async (request, reply) => {
    try {
      const payload = scriptCreateSchema.parse(request.body);
      const script = await prisma.$transaction(async (tx) => {
        const created = await tx.script.create({
          data: {
            name: payload.name,
            content: payload.content,
            source_type: payload.source_type
          }
        });
        await tx.scriptVersion.create({
          data: {
            script_id: created.id,
            version_no: 1,
            content: created.content
          }
        });
        return created;
      });
      return reply.code(201).send(
        serializeScriptRead({
          ...script,
          scene_count: 0,
          scenes: []
        })
      );
    } catch (error) {
      return sendError(reply, error);
    }
  });

  app.put("/api/scripts/:scriptId", async (request, reply) => {
    try {
      const scriptId = Number((request.params as { scriptId: string }).scriptId);
      const payload = scriptUpdateSchema.parse(request.body);
      const existing = await prisma.script.findUnique({ where: { id: scriptId } });
      if (!existing) {
        notFound("Script not found");
      }
      const script = await prisma.$transaction(async (tx) => {
        const updated = await tx.script.update({
          where: { id: scriptId },
          data: {
            name: payload.name ?? existing.name,
            content: payload.content ?? existing.content,
            source_type: payload.source_type ?? existing.source_type
          }
        });
        const currentVersion = await tx.scriptVersion.aggregate({
          where: { script_id: scriptId },
          _max: { version_no: true }
        });
        await tx.scriptVersion.create({
          data: {
            script_id: updated.id,
            version_no: (currentVersion._max.version_no ?? 0) + 1,
            content: updated.content
          }
        });
        return updated;
      });
      const [sceneCount, scenes] = await Promise.all([
        prisma.sceneScript.count({ where: { script_id: scriptId } }),
        scriptSceneReferences(scriptId)
      ]);
      return serializeScriptRead({
        ...script,
        scene_count: sceneCount,
        scenes
      });
    } catch (error) {
      return sendError(reply, error);
    }
  });

  app.delete("/api/scripts/:scriptId", async (request, reply) => {
    try {
      const scriptId = Number((request.params as { scriptId: string }).scriptId);
      const existing = await prisma.script.findUnique({ where: { id: scriptId } });
      if (!existing) {
        notFound("Script not found");
      }
      await prisma.script.delete({ where: { id: scriptId } });
      return reply.code(204).send();
    } catch (error) {
      return sendError(reply, error);
    }
  });

  app.post("/api/scripts/:scriptId/copy", async (request, reply) => {
    try {
      const scriptId = Number((request.params as { scriptId: string }).scriptId);
      const existing = await prisma.script.findUnique({ where: { id: scriptId } });
      if (!existing) {
        notFound("Script not found");
      }
      const copied = await prisma.$transaction(async (tx) => {
        const script = await tx.script.create({
          data: {
            name: `${existing.name}-copy`,
            content: existing.content,
            source_type: existing.source_type
          }
        });
        await tx.scriptVersion.create({
          data: {
            script_id: script.id,
            version_no: 1,
            content: script.content
          }
        });
        return script;
      });
      return reply.code(201).send(
        serializeScriptRead({
          ...copied,
          scene_count: 0,
          scenes: []
        })
      );
    } catch (error) {
      return sendError(reply, error);
    }
  });

  // 单脚本直接执行：内存编译脚本为 YAML 后调 Midscene 执行，不创建临时 scene
  app.post("/api/scripts/:scriptId/execute", async (request, reply) => {
    try {
      const scriptId = Number((request.params as { scriptId: string }).scriptId);
      const script = await prisma.script.findUnique({ where: { id: scriptId } });
      if (!script) {
        notFound("Script not found");
      }
      // 解析脚本内容为环境段 + task 快照，再合成可执行 YAML
      const scriptEnv = extractScriptEnv(script.content);
      // 纯 UI app 脚本（无平台段、无 http interface 配置）默认按安卓执行，自动补 android 段
      const hasPlatformEnv = ["android", "ios", "web", "computer"].some(
        (key) => scriptEnv[key] !== undefined
      );
      const hasHttpInterface = /^\s*interface\s*:/m.test(script.content);
      if (!hasPlatformEnv && !hasHttpInterface) {
        scriptEnv.android = scriptEnv.android ?? {};
      }
      const tasks = parseScriptTasks(script.content);
      const taskSnapshots = tasks.map((item) => dumpTaskSnapshot(item.task));
      const compiledYaml = compileSceneScript(scriptEnv, taskSnapshots);

      const execution = await executeCompiledScene({
        compiledYaml,
        taskSnapshots,
        httpTimeoutSec: env.SCENE_HTTP_TIMEOUT_SEC,
        defaultAndroidDeviceId: env.MIDSCENE_ANDROID_DEVICE_ID
      });

      return {
        script_id: script.id,
        script_name: script.name,
        task_count: tasks.length,
        success: Boolean(execution.success),
        message: String(execution.message),
        outputs: execution.outputs,
        detail: {
          task_results: execution.task_results,
          result: execution.result ?? execution.outputs
        }
      };
    } catch (error) {
      if (error instanceof SceneCompileError) {
        return reply.code(400).send({ detail: error.message });
      }
      if (error instanceof AndroidExecutionError) {
        badRequest(error.message);
      }
      return sendError(reply, error);
    }
  });
}
